Improbable/Postgres/Improbable.Postgres.CSharpCodeGen/ComponentGenerator.cs

```diff
@@ -93,71 +93,41 @@ public static string InitializeDatabase(string tableName)
-                        break;
-                    case FieldType.Singular:
-                        sb.AppendLine($"reader.{Types.SchemaToReaderMethod[field.SingularType.Type.Primitive]}({ordinal}),");
-                        break;
-                    default:
-                        throw new ArgumentOutOfRangeException();
-                }
+                    FieldType.Option => $"reader.IsDBNull({ordinal}) ? null : ({FqnFieldType(type, field)}) reader.{Types.SchemaToReaderMethod[field.OptionType.InnerType.Primitive]}({ordinal}),",
+                    FieldType.Singular => $"reader.{Types.SchemaToReaderMethod[field.SingularType.Type.Primitive]}({ordinal}),",
+                    _ => throw new ArgumentOutOfRangeException()
+                };
+
+                sb.AppendLine(toAdd);
             }
 
             return sb.ToString().TrimEnd().TrimEnd(',');
         }
 
         private static string PostgresTypeConversion(FieldDefinition f)
         {
-            switch (f.TypeSelector)
+            return f.TypeSelector switch
             {
-                case FieldType.Option:
-                    switch (f.OptionType.InnerType.ValueTypeSelector)
-                    {
-                        case ValueType.Primitive:
-                            if (f.OptionType.InnerType.Primitive == PrimitiveType.String)
-                            {
-                                return "::text";
-                            }
-
-                            break;
-                        default:
-                            throw new InvalidOperationException("Don't know how to convert this type from Postgres");
-                    }
-
-                    break;
-                case FieldType.Singular:
-                    switch (f.SingularType.Type.ValueTypeSelector)
-                    {
-                        case ValueType.Primitive:
-                            if (f.SingularType.Type.Primitive == PrimitiveType.String)
-                            {
-                                return "::text";
-                            }
-
-                            break;
-                        default:
-                            throw new InvalidOperationException("Don't know how to convert this type from Postgres");
-                    }
-
-                    break;
-                default:
-                    return string.Empty;
-            }
-
-            return string.Empty;
+                FieldType.Option when f.HasPrimitive(PrimitiveType.String) => "::text",
+                FieldType.Option when f.HasPrimitive() => string.Empty,
+                FieldType.Option => throw new InvalidOperationException($"Unsupported schema type '{f.OptionType.InnerType.Type}'. Only primitive types can be converted from Postgres."),
```

Comment:
This can probably be combined with the same error below using a fallthrough?

Reply:
Ooops, it actually can't because it's Option vs Singular
You found a bug :)